### PR TITLE
Fix precision loss in splash attention dsinks gradient (fixes #40128)

### LINKED ISSUE #40128: splash attention: `dsinks` gradient loses precision by rounding before the reduction
### Description

In the splash-attention backward pass, the gradient with respect to attention sinks
(`dsinks`) rounds its intermediate product to the low-precision input dtype before
summing, while `dq`/`dk`/`dv` in the very same function accumulate in `float32` and cast
only the final result. With `bfloat16` inputs this costs `dsinks` one to two orders of
magnitude of accuracy.

**Why `dsinks` is unusually sensitive.** It is not just "one more matmul". It is a
*fully reduced, heavily cancelling* sum: every element of `[num_heads, seq_len, head_dim]`
collapses to one scalar per head, and the positive and negative terms almost entirely
cancel. Measured below, `|sum| / sum|terms|` is `1e-5 … 6e-4` — the true result is four
to five orders of magnitude smaller than the mass being summed. In that regime the
`~0.4%` relative rounding that `bfloat16` applies *per term* no longer averages out: the
errors survive while the signal shrinks, so the relative error of the total explodes.
Casting after the reduction avoids this entirely, which is exactly what the neighbouring
gradients already do.

This also makes the kernel inconsistent with itself. Note that on `main` `do` has already
been promoted to `float32` on the line above, so `.astype(o.dtype)` actively throws that
promotion away:

```python
do = do.astype(jnp.float32)                                                   # promoted
dv = jnp.einsum("pt,pd->td", p, do).astype(v.dtype)                           # fp32 -> cast at end
dk = jnp.einsum("sd,st->td", q.astype(jnp.float32), ds).astype(k.dtype)       # fp32 -> cast at end
dq = jnp.einsum("st,td->sd", ds, k.astype(jnp.float32)).astype(q.dtype)       # fp32 -> cast at end
dsinks = jnp.sum(sinks_exp.astype(o.dtype) * do * o)                          # cast FIRST  <-- here
```

Because sinks are trainable parameters, this error is not a one-off: it is injected into
the sink update at every step of training.

### Affected code

Both backward paths, on `main` as of this report:

- `jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_kernel.py:319`
  (reference bwd, `_attention_reference_custom_bwd`)
- `jax/experimental/pallas/ops/tpu/splash_attention/splash_attention_kernel.py:2360`
  (`_splash_attention_bwd`)

### Reproduction

Pure JAX, no TPU required (`bfloat16` inputs, `float32` reference):

```python
import jax, jax.numpy as jnp, numpy as np

def _w(sinks, lse):
    return -jnp.exp(sinks[..., None, None].astype(jnp.float32)
                    - lse[..., None].astype(jnp.float32))

def current(sinks, lse, o, do):      # what jax does today
    return jnp.sum(_w(sinks, lse).astype(o.dtype) * o * do, axis=(-1, -2))

def proposed(sinks, lse, o, do):     # fp32 product, cast the result
    return jnp.sum(_w(sinks, lse) * o.astype(jnp.float32) * do.astype(jnp.float32),
                   axis=(-1, -2)).astype(o.dtype)

def exact(sinks, lse, o, do):        # fp32 reference
    w = jnp.exp(sinks[..., None].astype(jnp.float32) - lse.astype(jnp.float32))
    di = jnp.einsum("hsd,hsd->hs", o.astype(jnp.float32), do.astype(jnp.float32))
    return -jnp.sum(w * di, axis=-1)

H, D, S = 4, 128, 8192
rng = np.random.default_rng(0)
sinks = jnp.asarray(rng.normal(0, 1, (H,)), jnp.float32)
lse = jnp.full((H, S), np.log(S), jnp.float32)
o  = jnp.asarray(rng.normal(0, 1, (H, S, D)), jnp.bfloat16)
do = jnp.asarray(rng.normal(0, 1, (H, S, D)), jnp.bfloat16)

ref, cur, pro = exact(sinks, lse, o, do), current(sinks, lse, o, do), proposed(sinks, lse, o, do)
for h in range(H):
    r, c, p = float(ref[h]), float(cur[h]), float(pro[h])
    print(f"head {h}: ref={r:+.6e} current={c:+.6e} ({abs(c-r)/abs(r):.1%}) "
          f"proposed={p:+.6e} ({abs(p-r)/abs(r):.1%})")
```

Output (jax 0.8.3, CPU):

```
head 0: ref=-5.589731e-02 current=-5.590820e-02 (0.0%) proposed=-5.590820e-02 (0.0%)
head 1: ref=-4.731433e-02 current=-4.711914e-02 (0.4%) proposed=-4.736328e-02 (0.1%)
head 2: ref=+2.237111e-03 current=+2.838135e-03 (26.9%) proposed=+2.243042e-03 (0.3%)
head 3: ref=+5.379596e-02 current=+5.273438e-02 (2.0%) proposed=+5.371094e-02 (0.2%)
```

Head 2 is the worst case and it is the head with the most cancellation
(`|sum|/sum|terms| = 1.4e-05`), which is the mechanism described above. Per-head
cancellation ratios for this input are `6.0e-04, 6.6e-04, 1.4e-05, 5.9e-04`, over
`1,048,576` terms summed per head.

### Proposed fix

Keep the reduction in `float32` and cast only the result, matching `dq`/`dk`/`dv`:

```diff
-    dsinks = jnp.sum(sinks_exp.astype(o.dtype) * do * o)
+    dsinks = jnp.sum(sinks_exp * do.astype(jnp.float32)
+                     * o.astype(jnp.float32)).astype(o.dtype)
```

and in `_splash_attention_bwd`:

```diff
-    dsinks = jnp.sum(sinks_exp.astype(o.dtype) * o * do, axis=(-1, -2))
+    dsinks = jnp.sum(sinks_exp * o.astype(jnp.float32) * do.astype(jnp.float32),
+                     axis=(-1, -2)).astype(o.dtype)
```

The trailing `.astype(o.dtype)` is deliberate: it keeps the returned cotangent's dtype
unchanged, so this is a pure precision fix with no API or dtype-contract change. Dropping
it would leak `float32` cotangents for `bfloat16` sinks. The cost is negligible —
`dsinks` reduces to one scalar per head, and `do` is already `float32` at that point in
`_attention_reference_custom_bwd`.

### System info

```
jax 0.11.0, jaxlib 0.11.0
```
Verified against `main` (the two lines cited above are unchanged there).

### System info (python version, jaxlib version, accelerator, etc.)

```
jax 0.11.0
```

### PR DESCRIPTION
**Pull request by Aayush2492**
Fix for #40128

In the splash attention backward pass, dsinks previously truncated intermediate terms to o.dtype (e.g. bfloat16) before performing the summation / reduction. Because attention sink gradients involve heavily cancelling terms, early rounding led to substantial numerical cancellation and precision loss.

Adds regression test in tests/pallas/tpu_splash_attention_kernel_test.py.

### CODE REVIEW

**Comment by Aayush2492** [comment]
@sharadmv  @rdyro @Rifur13 Friendly ping for review when you have a moment.
